Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php

```diff
@@ -0,0 +1,259 @@
+use Neos\EventStore\Model\EventStream\VirtualStreamName;
+
+/**
+ * Content Repository service to perform migrations of events.
+ *
+ * Each function is used here for a specific migration. The migrations are only useful for production
+ * workloads which have events prior to the code change.
+ *
+ * @internal this is currently only used by the {@see MigrateEventsCommandController}
+ */
+final class EventMigrationService implements ContentRepositoryServiceInterface
+{
+    private bool $eventsTableWasUpdated = false;
+
+    public function __construct(
+        private readonly ContentRepositoryId $contentRepositoryId,
+        private readonly ContentRepository $contentRepository,
+        private readonly EventStoreInterface $eventStore,
+        private readonly Connection $connection,
+    ) {
+    }
+
+    /**
+     * The following things have to be migrated:
+     *
+     * - `NodePropertiesWereSet` `payload`
+     * - `SetSerializedNodeProperties` in `metadata`
+     *
+     * example:
+     *     "propertyValues":{"tagName":{"value":null,"type":"string"}}
+     * ->
+     *     "propertyValues":[],"propertiesToUnset":["tagName"]
+     *
+     * Here we just have to omit the null values because its a new node either way:
+     *
+     * - `NodeAggregateWithNodeWasCreated` `payload`
+     * - `CreateNodeAggregateWithNodeAndSerializedProperties` in `metadata`
+     *
+     * example:
+     *     "initialPropertyValues":{"title":{"value":"Blog","type":"string"},"titleOverride":{"value":null,"type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
+     * ->
+     *     "initialPropertyValues":{"title":{"value":"Blog","type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
+     *
+     * - `CopyNodesRecursively` in `metadata`
+     *
+     * example (somewhere deep recursive in $nodeTreeToInsert):
+     *     "propertyValues":{"senderName":{"value":null,"type":"string"}}}
+     * ->
+     *     "propertyValues":{}
+     *
+     * Needed for #4322: https://github.com/neos/neos-development-collection/pull/4322
+     *
+     * Included in February 2023 - before final Neos 9.0 release
+     *
+     * @param \Closure $outputFn
+     * @return void
+     */
+    public function migratePropertiesToUnset(\Closure $outputFn)
+    {
+        $this->eventsTableWasUpdated = false;
+
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId)
+            . '_bak_' . date('Y_m_d_H_i_s');
+        $outputFn(sprintf('Backup: copying events table to %s', $backupEventTableName));
+
+        $this->copyEventTable($backupEventTableName);
+
+        $streamName = VirtualStreamName::all();
+        $eventStream = $this->eventStore->load($streamName);
+        foreach ($eventStream as $eventEnvelope) {
+            $outputRewriteNotice = fn(string $message) => $outputFn(sprintf('%s@%s %s', $eventEnvelope->event->type->value, $eventEnvelope->sequenceNumber->value, $message));
```

Comment:
Maybe we can reduce the output. If one has many events warnings or migration messages are hard to find 😅

Reply:
yes i wasnt aware of 100k nodes :O Sorry to your terminal